documentation/projects/proposals/data_normalization/20240227-implementation_plan_catalog_data_cleaning.md

```diff
@@ -0,0 +1,198 @@
+
+### Save TSV files of cleaned data to AWS S3
+
+In a previous exploration, it was set to store TSV files of the cleaned data in
+the form of `<identifier> <cleaned_field>`, which can be used later to perform
+the updates efficiently in the catalog DB, which only had indexes for the
+`identifier` field. These files are saved to the disk of the Ingestion Server
+EC2 instances, and worked fine for files with URL corrections since this type of
+fields is relatively short, but became a problem when trying to save tags, as
+the file turned too large and filled up the disk, causing problems to the data
+refresh execution.
+
+The alternative is to upload TSV files to the Amazon Simple Storage Service
+(S3), creating a new bucket or using `openverse-catalog` with a subfolder. The
+benefit of using S3 buckets is that they have streaming capabilities and will
+allow us to read the files in chunks later if necessary for performance. The
+downside is that objects in S3 don't allow appending, so it may require to
+upload files with different part numbers or evaluate if the [multipart upload
+process][aws_mpu] will serve us here.
+
+[aws_mpu]:
+  https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html
+
+| timestamp (UTC)     | 'url' | 'creator_url' | 'foreign_landing_url' | 'tags' |
+| ------------------- | :---: | :-----------: | :-------------------: | :----: |
+| 2024-02-27 04:05:26 | 22156 |    9035458    |        8809213        |   0    |
+| 2024-02-20 04:06:56 | 22157 |    9035456    |        8809209        |   0    |
+| 2024-02-13 04:41:22 | 22155 |    9035451    |        8809204        |   0    |
+
+To have some numbers of the problem we are delaing with, the previous table
+shows the number of records cleaned by field for last runs at the moment of
+writing this IP, except for tags, which we don't have accurate registries since
+file saving was disabled.
+
+### Make and run a batched update DAG for one-time cleanup
+
+A batched catalog cleaner DAG (or potentially a `batched_update_from_file`)
+should take the files of the previous step to perform an batched update on the
+catalog's image table, while handling deadlocking and timeout concerns, similar
+to the [batched_update][batched_update]. This table is constantly in use by
+other DAGs, such as those from API providers or the data refresh process, and
+ideally can't be singly blocked by any DAG.
+
+[batched_update]: ./../../../catalog/reference/DAGs.md#batched_update
+
+A [proof of concept PR](https://github.com/WordPress/openverse/pull/3601)
+consisted of uploading each file to temporary `UNLOGGED` DB tables (which
+provides huge gains in writing performance while their disadventages are not
+relevant since they won't be permanent), and including a `row_id` serial number
+used later to query it in batches. Adding an index in this last column after
+filling up the table could improve the query performance. An adaptation will be
+needed to handle the column type of tags (`jsonb`).
```

Comment:
Do you mind adding a line at the end here explicitly noting that we'll need to run the DAGs on the previously generated TSVs as part of this step?

Reply:
It's mentioned in the first line of this section 😶 

> A batched catalog cleaner DAG (or potentially a `batched_update_from_file`) **should take the files of the previous step** to...

Does it need to be rephrased or moved?